plutus-tx/src/PlutusTx/AsData.hs

```diff
@@ -0,0 +1,148 @@
+pattern synonyms that match the behaviour of the original datatype.
+
+Since 'BuiltinData' can only contain 'BuiltinData', the pattern synonyms
+encode and decode all the fields using 'toBuiltinData' and 'unsafeFromBuiltinData'.
+That means that these operations are called on every pattern match or construction.
+This *can* be very efficient if, for example, the datatypes for the fields have
+also been defined with 'asData', and so have trivial conversions to/from 'BuiltinData'
+(or have very cheap conversions, like 'Integer' and 'ByteString').
+But it can be very expensive otherwise, so take care.
+
+Instances for 'ToData', 'FromData', and 'UnsafeFromBuiltinData' are automatically
+derived. Other instances can be derived using standalone deriving, and should mostly
+use the newtype strategy (e.g. in order to benefit from the fast 'Eq' instance for
+the underlying 'BuiltinData').
+
+Example:
+@
+    $(asData
+      [d|
+        data Example a = Ex1 Integer | Ex2 a a
+      |]
+@
+becomes
+@
+    newtype Example a = Example BuiltinData
+      deriving newtype (ToData, FromData, UnsafeFromBuiltinData)
+
+    pattern Ex1 :: (ToData a, UnsafeFromData a) => Integer -> Example a
+    pattern Ex1 i = Example (unsafeDataAsConstr -> ((==) 0 -> True, [unsafeFromBuiltinData -> i]))
+      where Ex1 i = Example (mkConstr 0 [toBuiltinData i])
+
+    pattern Ex2 :: (ToData a, UnsafeFromData a) => a -> a -> Example a
+    pattern Ex2 a1 a2 = Example (unsafeDataAsConstr -> ((==) 1 -> True, [unsafeFromBuiltinData -> a1, unsafeFromBuiltinData -> a2]))
+      where Ex2 a1 a2 = Example (mkConstr 1 [toBuiltinData a1, toBuiltinData a2])
+
+    {-# COMPLETE Ex1, Ex2 #-}
+@
+-}
+asData :: TH.Q [TH.Dec] -> TH.Q [TH.Dec]
+asData decQ = do
+  decs <- decQ
+  outputDecs <- for decs asDataFor
+  pure $ concat outputDecs
+
+asDataFor :: TH.Dec -> TH.Q [TH.Dec]
+asDataFor dec = do
+  -- th-abstraction doesn't include deriving clauses, so we have to handle that hereth-abstraction doesn't include deriving clauses, so we have to handle that here
+  case dec of
+    TH.DataD _ _ _ _ _ deriv | not (null deriv) -> fail "asData: can't handle deriving clauses"
+    _                                           -> pure ()
+
+  di@(TH.DatatypeInfo{TH.datatypeVariant=dVariant, TH.datatypeCons=cons, TH.datatypeName=name, TH.datatypeVars=tTypeVars}) <- TH.normalizeDec dec
+
+  -- Other stuff is data families and so on
+  unless (dVariant == TH.Datatype) $ fail $ "asData: can't handle datatype variant " ++ show dVariant
+  -- a fresh name for the new datatype, but same lexically as the old one
+  cname <- TH.newName (show name)
+  -- The newtype declaration
+  let ntD =
+        let con = TH.NormalC cname [(TH.Bang TH.NoSourceUnpackedness TH.NoSourceStrictness, TH.ConT ''BuiltinData)]
+            ntDeriv ctx = TH.DerivClause (Just TH.NewtypeStrategy) ctx
+            deriv = ntDeriv [TH.ConT ''ToData, TH.ConT ''FromData, TH.ConT ''UnsafeFromData]
```

Comment:
Why not derive other instances, particularly `Eq`, automatically?

Reply:
Just not being opinionated. For example, sometimes people don't _want_ an `Eq` instance. The job of this function is to define the datatype, users should still have full control over what instances they want or don't want.

I was going to say that the To/FromData instances are necessary, but that's not true. They're only necessary if you use the type in another AsData type, which you might not do. So we shouldn't provide those either.